pySDC/implementations/hooks/log_errors.py

```diff
@@ -24,14 +34,47 @@ def post_step(self, step, level_number):
             level=L.level_index,
             iter=step.status.iter,
             sweep=L.status.sweep,
-            type='e_global',
+            type=f'e_global{suffix}',
             value=abs(L.prob.u_exact(t=L.time + L.dt) - L.uend),
         )
 
+    def log_local_error(self, step, level_number, suffix=''):
+        """
+        Function to add the local error to the stats
+
+        Args:
+            step (pySDC.Step.step): The current step
+            level_number (int): The index of the level
+            suffix (str): Suffix for naming the variable in stats
+
+        Returns:
+            None
+        """
+        L = step.levels[level_number]
+
+        L.sweep.compute_end_point()
+
+        self.add_to_stats(
+            process=step.status.slot,
+            time=L.time + L.dt,
+            level=L.level_index,
+            iter=step.status.iter,
+            sweep=L.status.sweep,
+            type=f'e_local{suffix}',
+            value=abs(L.prob.u_exact(t=L.time + L.dt, u_init=L.u[0], t_init=L.time) - L.uend),
+        )
+
+
+class LogGlobalErrorPostStep(LogError):
+    def post_step(self, step, level_number):
+        super().post_step(step, level_number)
+        self.log_global_error(step, level_number, '_post_step')
+
 
 class LogGlobalErrorPostRun(hooks):
     """
     Compute the global error once after the run is finished.
+    Because of some timing issues, we cannot inherit from the `LogError` class here.
```

Comment:
Huh? What's the issue here?

Reply:
The problem is that the convergence controllers which change the step size are called after the `post_step` hooks but before the `post_run` hook. That gives the convergence controllers a chance to change the step size before recording the error at `t=L.time + L.dt` in the `post_run` hook, which will then be a different time than belongs to the solution, which was recorded at `t=L.time + L.dt` as well, but `L.dt` had a different value.
To take care of this, I record the time at which the solution was recorded in the `LogGlobalErrorPostRun` hook. I can make the evaluation time a parameter in the function that computes the error, but since there is not a huge amount of abstraction, I thought I just don't abstract here.

Reply:
Gives me headaches.. but ok.

Reply:
It was a weird bug that surprised me a little, but it is just a result of abstraction of the convergence controllers in the end. Why should they care what the controller decides to do next?
That is why the Lorenz test checks if the hook does what it's supposed to. I originally wanted to remove the computation from there and only use the hook, but then the bug showed itself ;)

Reply:
All hail automated tests!